### DISCUSSION TITLE
Markdown heading HTML IDs (or better `text` / `id` implementation)

### DISCUSSION THREAD

**Discussion by matthewlarkin** [sqlpage/SQLPage#203] [Q&A] [answered]
Dear SQLPagers,

I'm wondering if there's a way to add HTML ids to markdown headings, etc -- particularly helpful for inner document links (ie, _mysite.xyz#my_subheading_)

For instance, one might do the following:
```md
<!-- traditional approach -->
<h2 id='my_subheading'>My subheading here</h2>

<!-- some markdown parsers support this -->
## My subheading {#subheading}
```

SQLPage doesn't seem to support either as the traditional tag approach escapes the HTML characters, and the custom syntax is not supported by the SQLPage markdown parser.

- - - - -

**NOTE**: I initially thought the [`text`](https://sql.ophir.dev/documentation.sql?component=text#component) component would help here with it's `id` parameter, but it seems when you pass markdown to `contents_md`, the markdown gets parsed _first_ before the `text` block is rendered. For instance:
```sql
select 'text' as component,
  'about' as id,
  sqlpage.read_file_as_text('info/about.md') as contents_md;
```
renders something like the following:
```html
<!-- markdown gets rendered as HTML first -->
<h2>My subheading</h2>
<p>My content . . .</p>
<p id="about"></p> <!-- then SQLPage kicks in with the `text` container -->
```

- - - - -

☝️ That being said, there is a workaround: _calling the `text` component with the `id` first_ then for the first row selecting the markdown and passing it into the `contents_md` row level parameter ([_undocumented_](https://sql.ophir.dev/documentation.sql?component=text#component)).
```sql
select 'text' as component,
    'about' as id;
select sqlpage.read_file_as_text('assets/info/about.md') as contents_md;
```
This puts a blank paragraph with the `id` first, followed immediately by the parsed markdown.
```html
<p id="about"></p> <!-- SQLPage kicks in with the `text` container first -->
<!-- markdown gets rendered as HTML subsequently -->
<h2>My subheading</h2>
<p>My content . . .</p>
```
This does work, albeit a little hacky (as the markdown content is not given the id -- some rando paragraph prior gets it 😄). Although the paragraph does provide some nice margin on top the content when you do scroll to it ...👏

- - - - -

Still, I think allowing markdown headings IDs could be a more elegant solution. Any ideas? 🤔

**Comment by lovasoa** [comment, marked as answer]
The id coming after the rendered markdown was an oversight. This should be fixed. Can you test the latest build ?

**Reply by matthewlarkin** [reply]
Oh wonderful! I will check

**Comment by lovasoa** [comment]
Concerning new markdown syntax options, you can propose them / open pull requests on https://github.com/wooorm/markdown-rs

**Reply by matthewlarkin** [reply]
I'm not a Rust developer (yet), but I will touch base with a Rust colleague of mine who I've recently turned on to SQLPage 😎👌
